xarray/core/dataset.py

```diff
@@ -9023,7 +9023,10 @@ def polyfit(
         variables = {}
         skipna_da = skipna
 
-        x = get_clean_interp_index(self, dim, strict=False)
+        x = get_clean_interp_index(self, dim, use_coordinate=dim, strict=False)
```

Comment:
Agreed that it is strange, in an ideal world I think the `dim` argument could be either a coordinate or a dimension, and `use_coordinate` just wouldn't be necessary. But I didn't want to change the `get_clean_interp_index` signature too much since it seemed to match the public `interpolate_na` API quite closely, which I'm not familiar with.

I also didn't set the `dim` argument to `None` because there seemed to be some downstream use of that variable in the case where the resulting index is a DatetimeIndex.

Maybe as a half measure I could update the documentation of `get_clean_interp_index` in this PR to match more closely what the code is actually doing?

Alternatively, I could just inline the behaviour of `get_clean_interp_index` that we're actually using since that function seemed to be designed for a different purpose anyway.

Reply:
> I could just inline the behaviour of get_clean_interp_index that we're actually using since that function seemed to be designed for a different purpose anyway.

This would be great if it's a simple add

Reply:
Done :) although mypy complained at me initially for reassigning `x` to different types. I'm not sure how precise this project wants all its types to be so for now I've just declared `x` as `Any` for the sake of brevity.

Reply:
FYI: I am currently working on improvements of `interpolate_na` and discovered similar issues with `get_clean_interp_index`. I tried to come up with an improved (and type safe) version of it ([see here](https://github.com/pydata/xarray/blob/f19f626795512711f468a2fb898b12842491fbe9/xarray/core/missing.py#L388C1-L490)). However, good to know that polyfit is now independent, in this case `get_clean_interp_index` is only used within missing.py. I will try to write a better docstring for the function before merging.